File: server/src/utils/jwt.ts
import jwt from "jsonwebtoken";

export interface TokenPayload {
  userId: number;
  userName: string;
  userRole: number;
  login?: string;
}

const MIN_SECRET_LENGTH = 32;
const JWT_SECRET_PLACEHOLDERS = new Set([
  "your-secret-access-key-here",
  "your-secret-refresh-key-here",
  "change_me",
  "secret",
]);

const assertJwtSecret = (value: string | undefined, name: string) => {
  if (!value) {
    throw new Error(`${name} is not set`);
  }

  if (value.length < MIN_SECRET_LENGTH) {
    throw new Error(`${name} must be at least ${MIN_SECRET_LENGTH} characters long`);
  }

  if (JWT_SECRET_PLACEHOLDERS.has(value)) {
    throw new Error(`${name} must not use a placeholder value`);
  }
};

export const assertJwtSecrets = () => {
  assertJwtSecret(process.env.JWT_ACCESS_SECRET, "JWT_ACCESS_SECRET");
  assertJwtSecret(process.env.JWT_REFRESH_SECRET, "JWT_REFRESH_SECRET");
};

export const generateTokens = (userId: number, userName: string, userRole: number = 1, login?: string) => {
  const payload: TokenPayload = { userId, userName, userRole, login };

  const accessToken = jwt.sign(payload, process.env.JWT_ACCESS_SECRET!, {
    expiresIn: "15m",
  });

  const refreshToken = jwt.sign(payload, process.env.JWT_REFRESH_SECRET!, {
    expiresIn: "30d",
  });

  return { accessToken, refreshToken };
};

export const verifyAccessToken = (token: string): TokenPayload | null => {
  try {
    return jwt.verify(token, process.env.JWT_ACCESS_SECRET!) as TokenPayload;
  } catch {
    return null;
  }
};

export const verifyRefreshToken = (token: string): TokenPayload | null => {
  try {
    return jwt.verify(token, process.env.JWT_REFRESH_SECRET!) as TokenPayload;
  } catch {
    return null;
  }
};

File: server-go/utils/jwt.go
package utils

import (
	"errors"
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type TokenPayload struct {
	UserID   int    `json:"userId"`
	UserName string `json:"userName"`
	UserRole int    `json:"userRole"`
	Login    string `json:"login"`
	jwt.RegisteredClaims
}

const MinSecretLength = 32

var jwtSecretPlaceholders = map[string]bool{
	"your-secret-access-key-here":  true,
	"your-secret-refresh-key-here": true,
	"change_me":                    true,
	"secret":                       true,
}

// AssertJwtSecret checks that the secret is secure
func AssertJwtSecret(value string, name string) error {
	if value == "" {
		return fmt.Errorf("%s environment variable is not set", name)
	}

	if len(value) < MinSecretLength {
		return fmt.Errorf("%s must be at least %d characters long", name, MinSecretLength)
	}

	if jwtSecretPlaceholders[value] {
		return fmt.Errorf("%s must not use a placeholder value", name)
	}

	return nil
}

// GenerateTokens creates access and refresh tokens for a user
func GenerateTokens(userId int, userName string, userRole int, login string, accessSecret, refreshSecret string) (string, string, error) {
	// Access Token (expires in 15 minutes)
	accessClaims := &TokenPayload{
		UserID:   userId,
		UserName: userName,
		UserRole: userRole,
		Login:    login,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(15 * time.Minute)),
		},
	}

	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, accessClaims)
	accessSigned, err := accessToken.SignedString([]byte(accessSecret))
	if err != nil {
		return "", "", err
	}

	// Refresh Token (expires in 30 days)
	refreshClaims := &TokenPayload{
		UserID:   userId,
		UserName: userName,
		UserRole: userRole,
		Login:    login,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(30 * 24 * time.Hour)),
		},
	}

	refreshToken := jwt.NewWithClaims(jwt.SigningMethodHS256, refreshClaims)
	refreshSigned, err := refreshToken.SignedString([]byte(refreshSecret))
	if err != nil {
		return "", "", err
	}

	return accessSigned, refreshSigned, nil
}

// VerifyToken parses and validates a JWT token using the provided secret
func VerifyToken(tokenString string, secret string) (*TokenPayload, error) {
	token, err := jwt.ParseWithClaims(tokenString, &TokenPayload{}, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", t.Header["alg"])
		}
		return []byte(secret), nil
	})

	if err != nil {
		return nil, err
	}

	if claims, ok := token.Claims.(*TokenPayload); ok && token.Valid {
		return claims, nil
	}

	return nil, errors.New("invalid token claims")
}
